include/amici_model.h

```diff
@@ -916,26 +769,193 @@ namespace amici {
+         in the backward problem
+         *
+         * @param rootsfound provides the direction of the zero-crossing, so adding
+         it will give the right update to the heaviside variables (zero if no root
+         was found)
+         */
+        void updateHeavisideB(const int *rootsfound) {
+            for (int ie = 0; ie < ne; ie++) {
+                h.at(ie) -= rootsfound[ie];
+            }
+        }
+        
+        void fw(const realtype t, const N_Vector x);
+        
+        /** model specific implementation of fw
+         * @param w Recurring terms in xdot
+         * @param t timepoint
+         * @param x Vector with the states
+         * @param p parameter vector
+         * @param k constants vector
+         * @param h heavyside vector
+         */
+        virtual void model_w(realtype *w, const realtype t, const realtype *x, const realtype *p,
+                                const realtype *k, const realtype *h) {};
+        
+        void fdwdp(const realtype t, const N_Vector x);
+        
+        /** model specific implementation of dwdp
+         * @param dwdp Recurring terms in xdot, parameter derivative
+         * @param t timepoint
+         * @param x Vector with the states
+         * @param p parameter vector
+         * @param k constants vector
+         * @param h heavyside vector
+         * @param w vector with helper variables
+         */
+        virtual void model_dwdp(realtype *dwdp, const realtype t, const realtype *x, const realtype *p,
+                                const realtype *k, const realtype *h, const realtype *w) {};
+        
+        void fdwdx(const realtype t, const N_Vector x);
+        
+        /** model specific implementation of dwdx
+         * @param dwdx Recurring terms in xdot, state derivative
+         * @param t timepoint
+         * @param x Vector with the states
+         * @param p parameter vector
+         * @param k constants vector
+         * @param h heavyside vector
+         * @param w vector with helper variables
+         */
+        virtual void model_dwdx(realtype *dwdx, const realtype t, const realtype *x, const realtype *p,
+                                const realtype *k, const realtype *h, const realtype *w) {};
+        
+    protected:
+        
+        void getmy(const int it, const ExpData *edata);
+        
+        void gety(const int it, const ReturnData *rdata);
+        
+        void getx(const int it, const ReturnData *rdata);
+        
+        void getsx(const int it, const ReturnData *rdata);
+        
+        const realtype gett(const int it, const ReturnData *rdata) const;
+        
+        void getmz(const int nroots, const ExpData *edata);
+        
+        void getz(const int nroots, const ReturnData *rdata);
+        
+        void getrz(const int nroots, const ReturnData *rdata);
+        
+        /** Jacobian */
+        SlsMat J = nullptr;
+        
+        /** current state */
+        std::vector<double> x;
```

Comment:
yes and no ReturnData only contains (and probably must?) doubles as it interfaces with matlab.  It would be nice if we could at some point use` long double` or `single` aswell. This functions reads from ReturnData (via getx), but those functions are probably the best place to do perform the conversion anyways.

Reply:
Hm, then we should draw a clear line between `realtype` and `double` usage.